src/controllers/legacycontrollersettingsfactory.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <QDomElement>
+#include <QLabel>
+#include <QLayout>
+#include <QList>
+#include <QObject>
+#include <QSizePolicy>
+#include <QSpinBox>
+#include <QString>
+#include <QWidget>
+#include <limits>
+#include <memory>
+#include <tuple>
+
+#include "controllers/legacycontrollersettingsfactory.h"
+
+class AbstractLegacyControllerSetting;
+
+/// @brief This class defines an interface that a controller setting type must
+/// implement so it can be used properly by the builder
+/// @tparam T The class implementing this interface
+template<class T>
+class LegacyControllerSettingFactory {
+    inline static LegacyControllerSettingFactory* createFrom(const QDomElement& element) {
+        return new T(element);
+    }
+    inline static bool match(const QDomElement& element) {
+        return T::match(element);
+    }
+};
+
+/// @brief This class is used to dynamically instantiate a controller setting based on its type
+class LegacyControllerSettingBuilder {
+  public:
+    static LegacyControllerSettingBuilder* instance();
+
+    /// @brief Register a new type of setting. This method is used by the
+    /// REGISTER macro, it shouldn't be used directly
+    /// @param match the match function of the new setting
+    /// @param creator the creator function of the new setting
+    /// @return Always true
+    bool registerType(bool (*match)(const QDomElement&),
+            AbstractLegacyControllerSetting* (*creator)(const QDomElement&)) {
+        m_supportedSettings.append(std::make_tuple(match, creator));
+        return true;
+    }
+
+    /// @brief instantiate a new setting from a an XML definition if any valid
+    /// setting was found. The caller is the owner of the instance
+    /// @param element The XML element to parse to build the new setting
+    /// @return an instance if a a supported setting has been found, null
+    /// otherwise
+    static AbstractLegacyControllerSetting* build(const QDomElement& element) {
+        for (const auto& settingType : qAsConst(instance()->m_supportedSettings)) {
+            if (std::get<0>(settingType)(element)) {
+                return std::get<1>(settingType)(element);
```

Comment:
This code is hard to understand. Can you in introduce a private struct for proper naming?

Reply:
Alternative Idea: join match and create. In case we have no match return nullptr from create and continue the loop.

Reply:
Currently the match functions are all similar. Do we expect special matching? 
If not m_supportedSettings vam become a QHash with the "typ" as key. This will also speed up this loop.

Reply:
The reason of introducing a separate `match` and `create` is to deal with overlapping or similar-looking setting type. Currently, the settings that this PR implement shouldn't suffer from this, but once we introduce [controller screen render](https://github.com/mixxxdj/mixxx/pull/11407), they may be more complex settings types such as color or file path (e.g to set a custom background), which will be using the same generic type.